### PR TITLE
New code for isosurface trace

### PR DESCRIPTION
**Pull request by archmoj**
Supersedes #3311 i.e. without using `gl-isosurface3d` refactored code!
Totally rewritten and based on plotly.js most recent source code for `mesh3d` trace. 
Also in this regard, the [isosurface](https://www.npmjs.com/package/isosurface) module which used to support various algorithms namely `marching-cube`, `marching-tetrahedra` & `surface-nets` for displaying mathematical functions is forked and improved to accept data cubes. 
Using `isovalue` attribute the user may create one, two or multiple 3D contour surfaces.
It is possible to enable `isocap` surfaces too.

Various iso-surface traces could be modelled using [this manual demo](https://codepen.io/MojtabaSamimi/pen/dwXaaz?editors=0010) or [this auto-update demo](https://codepen.io/MojtabaSamimi/pen/yGJWaG).

gl3d_isosurface_math

@plotly/plotly_js

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -70,6 +70,7 @@
     "es6-promise": "^3.0.2",
     "fast-isnumeric": "^1.1.2",
     "font-atlas-sdf": "^1.3.3",
+    "isosurface": "git://github.com/archmoj/isosurface.git#e2c41594417a2ba363bb430c7d38adfe18312fc5",
```
Comment (archmoj) [inline package.json]: https://github.com/mikolalysenko/isosurface/compare/master...archmoj:plotly-integration

**Comment by archmoj** [comment]
With iso-caps now implemented, this should be ready for the first round review.  
@etpinard @antoinerg @alexcjohnson

**Comment by jackparmer** [comment]
Looks amazing! 🥇

**Comment by archmoj** [comment]
Thanks @jackparmer for testing this & the medal.
On Friday afternoon I decided to rewrite this thing and base it on @mikolalysenko's [isosurface module](https://www.npmjs.com/package/isosurface) which is great for mathematical formulas!

**Comment by jackparmer** [comment]
cc @empet, who may be interested in this new trace type

**Comment by empet** [comment]
@jackparmer  Yes, I'm interested.

File: src/traces/isosurface/attributes.js [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+    },
+    z: {
+        valType: 'data_array',
+        role: 'info',
+        editType: 'calc+clearAxisTypes',
+        description: [
+            'Sets the Z coordinates of the vertices on Z axis.'
+        ].join(' ')
+    },
+    volume: {
```
Comment (etpinard) [inline src/traces/isosurface/attributes.js]: Why did you change `value` -> `volume`?

Comment (archmoj) [reply src/traces/isosurface/attributes.js]: We could use `value` there. But I liked your comment here: 
https://github.com/plotly/plotly.js/pull/2752#discussion_r201448179

File: package-lock.json [hunk 3]
```diff
@@ -6077,6 +6077,10 @@
       "integrity": "sha1-TkMekrEalzFjaqH5yNHMvP2reN8=",
       "dev": true
     },
+    "isosurface": {
+      "version": "git://github.com/archmoj/isosurface.git#e2c41594417a2ba363bb430c7d38adfe18312fc5",
```
Comment (etpinard) [inline package-lock.json]: We should move that code over to `gl-isosurface`. That way it will be under the `gl-vis` organisation which all of us have access to.

Comment (etpinard) [reply package-lock.json]: Oops, I didn't notice that `isosurface` repo is a fork of https://github.com/mikolalysenko/isosurface - we should try to get your branch merge into Mikola's official repo. My apologies.

File: src/traces/isosurface/attributes.js [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+        editType: 'calc',
+        description: [
+            'Sets the text elements associated with the vertices.',
+            'If trace `hoverinfo` contains a *text* flag and *hovertext* is not set,',
+            'these elements will be seen in the hover labels.'
+        ].join(' ')
+    },
+
+    // Color field
+    color: {
```
Comment (etpinard) [inline src/traces/isosurface/attributes.js]: Do we really need `color`, `vertexcolor` and `facecolor`? Should applying a colorscale to `value` (your `volume`) be enough?

File: src/traces/isosurface/attributes.js [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+        description: [
+            'Sets the color of each face',
+            'Overrides *color* and *vertexcolor*.'
+        ].join(' ')
+    },
+    transforms: undefined
+},
+
+colorscaleAttrs('', {
+    colorAttr: '`intensity`',
```
Comment (etpinard) [inline src/traces/isosurface/attributes.js]: Should be `'value'` here.

File: src/traces/isosurface/attributes.js [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+        role: 'style',
+        dflt: false,
+        editType: 'calc',
+        description: [
+            'Determines whether or not normal smoothing is applied to the isosurfaces,',
+            'creating isosurfaces with an angular, low-poly look via flat reflections.'
+        ].join(' ')
+    },
+
+    contour: {
```
Comment (etpinard) [inline src/traces/isosurface/attributes.js]: `contour` doesn't appear to have been implemented (which is fine :smile: ) - let's :hocho: these attributes

File: src/traces/isosurface/attributes.js [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+    isovalue: {
+        valType: 'data_array',
+        role: 'info',
+        editType: 'calc',
+        dflt: [0],
+        description: [
+            'Sets iso surface boundaries.'
+        ].join(' ')
+    },
+    meshalgo: {
```
Comment (etpinard) [inline src/traces/isosurface/attributes.js]: I'd vote for removing this attribute from first version. Using marching cubes appears sufficient for now - unless @archmoj can convince me otherwise.

File: src/traces/isosurface/attributes.js (hunk 6 above)
Comment (archmoj) [reply src/traces/isosurface/attributes.js]: The contours seem to be working on hover namely on caps.

File: src/traces/isosurface/attributes.js (hunk 2 above)
Comment (etpinard) [reply src/traces/isosurface/attributes.js]: Ah nice find. Thanks! Let's keep `volume` for now. I might change my mind again, and we could use other opinions.

File: src/traces/isosurface/convert.js [hunk 8]
```diff
@@ -0,0 +1,480 @@
...
+    data.intensity = allCs;
+
+    return data;
+}
+
+function createIsosurfaceTrace(scene, data) {
+
+    var gl = scene.glplot.gl;
+
+    var mesh = createMesh({gl: gl});
```
Comment (etpinard) [inline src/traces/isosurface/convert.js]: Reusing `gl-mesh3d` is a very nice idea. One drawback I noticed though: hover labels don't appear on the input data coordinates, they appear on the (interpolated) mesh x/y/z - which I think is _ok_ for now, albeit not consistent with our other trace types. For example, surface hover always picks out data-coordinates (even when the surface coordinates are refined). In contrast, [streamtubes](https://rreusser.github.io/plotly-mock-viewer/#gl3d_streamtube-thin) hover picks the coordinates on the tubes themselves.

File: src/traces/isosurface/attributes.js (hunk 2 above)
Comment (alexcjohnson) [reply src/traces/isosurface/attributes.js]: hmm, I don't necessarily read that [matlab doc page](https://www.mathworks.com/help/matlab/ref/isosurface.html) as calling it "volume", they say "volume data V" but I read that as "V is a 3D array" rather than "V is called volume". I think I prefer `value`; `volume` implies that it's an array of volumes, but really it's an array of the value of some field/function dispersed over a volume.

Comment (archmoj) [reply src/traces/isosurface/attributes.js]: OK. The `value` or `values` could be applied in the next revision. The previous isosurface & volume traces used both two variations! 
@alexcjohnson please let me know which one to pick?

File: src/traces/isosurface/defaults.js [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+    coerce('x');
+    coerce('y');
+    coerce('z');
+
+    var handleCalendarDefaults = Registry.getComponentMethod('calendars', 'handleTraceDefaults');
+    handleCalendarDefaults(traceIn, traceOut, ['x', 'y', 'z'], layout);
+
+    // Coerce remaining properties
+    [
+        'volume',
```
[the remark this answers is not in the document]
Comment (archmoj) [reply src/traces/isosurface/defaults.js]: Should we add similar check in place for the `mesh3d` trace?

File: src/traces/isosurface/attributes.js (hunk 2 above)
Comment (archmoj) [reply src/traces/isosurface/attributes.js]: The `value` is applied instead of `volume` and the mocks are codepen demos are updated in case @empet wanted to play with this trace...

File: src/traces/isosurface/attributes.js (hunk 5 above)
Comment (archmoj) [reply src/traces/isosurface/attributes.js]: This `intensity` array is filled internally by the `convert.js` and provided to `gl-mesh3d` module. So I thought we need to keep it this way.

**Comment by archmoj** [comment]
@etpinard This is [a demo] related to the comment that you mentioned: https://github.com/plotly/plotly.js/pull/2752#issuecomment-403971750

File: src/traces/isosurface/attributes.js (hunk 7 above)
Comment (archmoj) [reply src/traces/isosurface/attributes.js]: The `SurfaceNets` algorithm is capable of creating smoother isosurfaces. The only issue with that is the end points are not reaching the cube boundaries. Anyway that may be a good option for the user in particular scenarios. Example below presents the mock with `SurfaceNets`.
isosurface_using_surfacenets

**Comment by archmoj** [comment]
@etpinard demo [link](https://codepen.io/MojtabaSamimi/pen/QzdbEM) related to https://github.com/plotly/plotly.js/pull/2752#issuecomment-403971750

File: src/traces/isosurface/attributes.js (hunk 4 above)
Comment (archmoj) [reply src/traces/isosurface/attributes.js]: Reduced color attributes in https://github.com/plotly/plotly.js/pull/3340/commits/49aa7468d0ea7917e1917c49620c964332b1e0c2

File: src/traces/isosurface/attributes.js (hunk 5 above)
Comment (archmoj) [reply src/traces/isosurface/attributes.js]: Used `value` instead of `intensity` there: https://github.com/plotly/plotly.js/pull/3340/commits/d9c1fa7e88309d6e500d2d12c178ba8c77b367d5

File: src/traces/isosurface/attributes.js (hunk 7 above)
Comment (archmoj) [reply src/traces/isosurface/attributes.js]: Done in https://github.com/plotly/plotly.js/pull/3340/commits/dd08cd463f2f6f9ace4fbe5da44748fef21ad2f8 .
